src/engine/init.c

```diff
@@ -558,24 +558,40 @@ dss_crt_event_cb(d_rank_t rank, uint64_t incarnation, enum crt_event_source src,
 		 enum crt_event_type type, void *arg)
 {
 	int			 rc = 0;
-	struct engine_metrics	*metrics;
+	struct engine_metrics	*metrics = &dss_engine_metrics;
 
 	/* We only care about dead ranks for now */
-	if (src != CRT_EVS_SWIM || type != CRT_EVT_DEAD) {
-		D_DEBUG(DB_MGMT, "ignore src/type/evict %u/%u\n",
-			src, type);
+	if (type != CRT_EVT_DEAD) {
+		D_DEBUG(DB_MGMT, "ignore: src=%d type=%d\n", src, type);
 		return;
 	}
 
-	metrics = &dss_engine_metrics;
-
-	d_tm_inc_counter(metrics->dead_rank_events, 1);
 	d_tm_record_timestamp(metrics->last_event_time);
 
-	rc = ds_notify_swim_rank_dead(rank, incarnation);
-	if (rc)
-		D_ERROR("failed to handle %u/%u event: "DF_RC"\n",
-			src, type, DP_RC(rc));
+	if (src == CRT_EVS_SWIM) {
+		d_tm_inc_counter(metrics->dead_rank_events, 1);
+		rc = ds_notify_swim_rank_dead(rank, incarnation);
+		if (rc)
+			D_ERROR("failed to handle %u/%u event: "DF_RC"\n",
+				src, type, DP_RC(rc));
+	} else if (src == CRT_EVS_GRPMOD) {
+		d_rank_t self_rank = dss_self_rank();
+
+		if (rank == dss_self_rank()) {
+			D_WARN("raising SIGTERM: exclusion of this engine (rank %u) detected\n",
+			       self_rank);
+			/*
+			 * For now, we just raise a SIGTERM to ourselves; we could
+			 * inform daos_server, who would initiate a termination and
+			 * decide whether to restart us.
+			 */
+			rc = kill(getpid(), SIGTERM);
```

Comment:
I think it's fine to allow the engine to handle this internally rather than involving the control plane for a decision, at least for now. One suggestion would be to add a new RAS event that the engine sends prior to shutting itself down. That way the control plane can expect the shutdown and make a decision, or at least give the admin some additional context.

I wonder if SIGTERM is best, though. In theory, the engine could hang in signal handlers and potentially still cause problems because it's not completely shut down, right? Do you see any problem with a more abrupt SIGKILL termination?

Reply:
> I think it's fine to allow the engine to handle this internally rather than involving the control plane for a decision, at least for now. One suggestion would be to add a new RAS event that the engine sends prior to shutting itself down. That way the control plane can expect the shutdown and make a decision, or at least give the admin some additional context.

That's mostly what I have in mind too. The only difference is that I think the control plane might prefer to initiate the termination itself, so that the RAS event and the SIGCHLD (i.e., the "engine exited" event) would not be racing; the engine side doesn't care too much about this.

> I wonder if SIGTERM is best, though. In theory, the engine could hang in signal handlers and potentially still cause problems because it's not completely shut down, right? Do you see any problem with a more abrupt SIGKILL termination?

Sure, I'll try SIGKILL. To be honest, don't recall why I chose SIGTERM---it's been more than a year since.